dace/libraries/standard/nodes/reduce.py

```diff
@@ -1075,6 +1077,403 @@ def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG, partial_width=16):
+        outsubset = dcpy(outedge.data.subset)
+        osqdim = outsubset.squeeze()
+        # input_dims = len(insubset)
+        # output_dims = len(outsubset)
+        input_data = sdfg.arrays[inedge.data.data]
+        output_data = sdfg.arrays[outedge.data.data]
+
+        in_type = input_data.dtype
+
+        if input_data.storage != dtypes.StorageType.GPU_Global:
+            # data doesnt reside on GPU --> return pure expansion
+            return ExpandReducePure.expansion(node, state, sdfg)
+
+        if len(osqdim) == 0:  # Fix for scalars
+            osqdim = [0]
+
+        # Standardize and squeeze axes
+        axes = node.axes if node.axes else [i for i in range(len(inedge.data.subset))]
+        axes = [axis for axis in axes if axis in isqdim]
+
+        assert node.identity is not None
+
+        # call the planner script
+        schedule = get_reduction_schedule(input_data, axes)
+
+        if schedule['error']:
+            # return pure expansion if error
+
+            ########################################################
+            ###    This gives segfault if we reduce all axes     ###
+            ###(GPU map with 1 thread block doesn't seem to work)###
+            ########################################################
+
+            pure_sdfg = ExpandReducePure.expansion(node, state, sdfg)
+            # need to apply gpu transformations, else runs on CPU (and our input data is on GPU)
+            pure_sdfg.apply_gpu_transformations()
+            return pure_sdfg
+
+        # Create nested SDFG
+        nsdfg = SDFG('reduce')
+        nsdfg.add_array('_in',
+                        schedule['in_shape'],
+                        input_data.dtype,
+                        strides=schedule['in_strides'],
+                        storage=dace.StorageType.GPU_Global)
+        nsdfg.add_array('_out',
+                        schedule['out_shape'],
+                        output_data.dtype,
+                        strides=schedule['out_strides'],
+                        storage=dace.StorageType.GPU_Global)
+        nstate = nsdfg.add_state()
+
+        # Interleave input and output axes to match input memlet
+        ictr, octr = 0, 0
+        input_subset = []
+        dims = list(range(len(schedule['in_shape'])))
+        for i in dims:
+            if i in schedule['axes']:
+                input_subset.append('_i%d' % ictr)
+                ictr += 1
+            else:
+                input_subset.append('_o%d' % octr)
+                octr += 1
+
+        if (schedule['contiguous_dim']):
+            # we are reducing the contiguous dimension
+
+            vectorize = schedule['vectorize']
+
+            ome, omx = nstate.add_map('block_grid',
+                                      {'_o%d' % i: '0:%s' % symstr(sz)
+                                       for i, sz in enumerate(schedule['grid'])},
+                                      schedule=dtypes.ScheduleType.GPU_Device)
```

Comment:
You should probably check within `can_be_applied` if we already in a GPU kernel or not. Reductions can appear within other kernels after subgraph fusions